### PR TITLE
[WIP] Onset of Convection Benchmark 

### PR DESCRIPTION
**Pull request by maxrudolph**
This is the product of a classroom exercise to compare numerical predictions with the analytic result from the linear stability analysis for the onset of convection in a fluid layer heated from below that appears in Turcotte and Schubert section 6-19.

### CODE REVIEW

**Comment by bangerth** [comment]
This is interesting, but doesn't quite fit into the existing scheme of cookbooks. Are you planning to add a section to the manual that shows the results? Of course with reference to the iPython files -- which would be useful for others interested in playing with this.

**Comment by maxrudolph** [comment]
Agree that it doesn't really fit with cookbooks. It is a benchmark in some sense. Happy to add it as a section to the manual, if you think that it might be useful there.

**Comment by bangerth** [comment]
Yes. It doesn't have to be overly long, but since we collect everything in the manual, a page or two there would be great.

**Comment by tjhei** [comment]
I actually think that it fits into cookbooks. (my mental model is: "is it a test? if not, is it something well-explained in the manual, it goes into cookbooks/. Otherwise benchmark/")

**Comment by maxrudolph** [comment]
Where is the best place for this to go in the manual? Towards the beginning of section 5.3? The other possibility is to think of this as a benchmark and put the manual description in 5.4.

**Comment by bangerth** [comment]
On 02/04/2017 12:18 PM, Max Rudolph wrote:
> Where is the best place for this to go in the manual? Towards the beginning of
> section 5.3? The other possibility is to think of this as a benchmark and put
> the manual description in 5.4.

Since you're dealing with a model problem, I think section 5.2 may actually be 
a good place. Maybe after 5.2.3? The benchmark section 5.4 would also seem 
like a reasonable place.

**Comment by maxrudolph** [comment]
@bangerth @tjhei Could you take a look at this again and let me know whether you think it needs anything further?

File: benchmark/onset-of-convection/convection-box-base.prm [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# This is the 'base' input file from which additional input files are generated
+# for the onset of convection example. UPPERCASE strings are replaced automatically
+# with appropriate values for each model run. This .prm file will fail if run
```
Comment (bangerth) [inline benchmark/onset-of-convection/convection-box-base.prm]: can you say what/who does this automatic replacement?

File: doc/manual/manual.bib [hunk 2]
```diff
@@ -10711,6 +10711,13 @@ @Book{STO01
   publisher =    {Cambridge},
   year =         2001}
 
+@Book{TS14,
+  author =       {D.L. Turcotte and G. Schubert},
```
Comment (bangerth) [inline doc/manual/manual.bib]: Can you add a space between `D.L.`?

File: doc/manual/manual.tex [hunk 3]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
 \texttt{solcx.prm} refers to the plugin as \texttt{./libsolcx.so}, i.e., in
 the current directory.
 
+\subsubsection{Onset of convection benchmark}
+\label{sec:benchmark-onset-of-convection}
+\textit{This section contributed by Max Rudolph, based on a course assignment for ``Geodynamic Modeling'' at Portland State University.}
```
Comment (bangerth) [inline doc/manual/manual.tex]: -> WAS contributed

File: doc/manual/manual.tex [hunk 4]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
 
+Here we use ASPECT to numerically reproduce numerically the results of a linear stability analysis for the onset of convection in a fluid layer heated from below. This exercise was assigned to students at Portland State University as a first step towards setting up a nominally Earth-like mantle convection model. Hence, representative length scales and transport properties for Earth are used. This cookbook consists of a jupyter notebook (\texttt{benchmarks/onset-of-convection/onset-of-convection.ipynb}) that is used to run aspect and analyze the results of several calculations. To use this code, you must compile aspect and give the path to the executable in the notebook as \texttt{aspect\_bin}.
```
Comment (bangerth) [inline doc/manual/manual.tex]: Here and below, please use `\aspect{}` instead of ASPECT or aspect.

File: doc/manual/manual.tex [hunk 5]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+
+The linear stability analysis for the onset of convection appears in Turcotte and Schubert \cite{TS14} (section 6.19). The linear stability analysis assumes the Boussinesq approximation and makes predictions for the growth rate (vertical velocity) of instabilities and the critical Rayleigh number $Ra_{cr}$ above which convection will occur. $Ra_{cr}$ depends only on the dimensionless wavelength of the perturbation, which is assumed to be equal to width of the domain. The domain has height $b$ and width $\lambda$ and the perturbation is described by:
```
Comment (bangerth) [inline doc/manual/manual.tex]: -> equal to THE width of the domain

Drop the colon at the end of the line and remove the following empty line to make the formula a regular part of the sentence -- it should grammatically be treated in the same way as if it had been an inlined formula that is just a regular part of the sentence.

File: doc/manual/manual.tex [hunk 6]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+
+\begin{align*}
+T'(x,y) = T_0'\cos\left(\frac{2\pi x}{\lambda}\right)\sin\left(\frac{\pi y}{b} \right)
```
Comment (bangerth) [inline doc/manual/manual.tex]: Then end the formula with a period.

File: doc/manual/manual.tex [hunk 7]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+\end{align}
+
+Note that because we place the bottom boundary of the domain at $y=0$ and the top at $y=b$, the perturbation vanishes at the top and bottom boundaries. This departs slightly from the setup in \cite{TS14} where the top and bottom boundaries of the domain are at $y=\pm b/2$.
+
+The analytic expression for the critical Rayleigh number, $Ra_{cr}$ is given in Turcotte and Schubert \cite{TS14} equation 6.319:
```
Comment (bangerth) [inline doc/manual/manual.tex]: comma after `\cite{...}`.

In the book, are formulas parenthesized? If so, add the parentheses here as well.

File: doc/manual/manual.tex [hunk 8]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+
```
Comment (bangerth) [inline doc/manual/manual.tex]: no empty line -- you're not starting a new paragraph.

File: doc/manual/manual.tex [hunk 9]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+\begin{align*}
+Ra_{cr}=\frac{\left(\pi^2+\frac{4\pi^2 b^2}{\lambda^2}\right)^3}{\frac{4\pi^2 b^2}{\lambda^2}}
```
Comment (bangerth) [inline doc/manual/manual.tex]: and end the formula with a period

File: doc/manual/manual.tex [hunk 10]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+\end{align}
+
+The linear stability analysis also makes a prediction for the dimensionless growth rate of the instability $\alpha'$ (Turcotte and Schubert \cite{TS14} equation 6.315). The maximum vertical velocity is given by:
```
Comment (bangerth) [inline doc/manual/manual.tex]: same comments as above: comma after `\cite`, parentheses

File: doc/manual/manual.tex [hunk 11]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+
```
Comment (bangerth) [inline doc/manual/manual.tex]: no empty line, no colon on line 8191

File: doc/manual/manual.tex [hunk 12]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+\begin{align*}
+v_{y,max} = \frac{2\pi}{\lambda}\phi_0' e^{\alpha' t},
+\end{align}
+
+where
+
```
Comment (bangerth) [inline doc/manual/manual.tex]: no empty line in 8196 and 8198

File: doc/manual/manual.tex [hunk 13]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+\begin{align*}
+\phi_0' = -\frac{2\pi}{\lambda}\frac{\rho_0 g \alpha T_0'}{\mu}\left(\frac{4\pi^2}{\lambda^2}+\frac{\pi^2}{b^2} \right)^{-2},
+\end{align}
+
+and
+
```
Comment (bangerth) [inline doc/manual/manual.tex]: same here

File: doc/manual/manual.tex [hunk 14]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+\begin{align*}
+\alpha'=\frac{\kappa}{b^2}\left[\frac{\rho_0 g \alpha b^3 \Delta T}{\mu \kappa}\left(\frac{\frac{4\pi^2 b^2}{\lambda^2}}{\left(\frac{4\pi^2 b^2}{\lambda^2}+\pi^2\right)^2}\right) -\left(\pi^2+\frac{4\pi^2b^2}{\lambda^2}\right)\right].
+\end{align}
+
+We use bisection to determine $Ra_{cr}$ for specific choices of the domain geometry, keeping the depth $b$ constant and varying the domain width $\lambda$. If the vertical velocity increases from the first to the second timetsp, the system is unstable to convection. Otherwise, it is stable. Each calculation is terminated after the second timestep. Fig. \ref{fig:onset-1} shows the numerically-determined threshold for the onset of convection, which can be compared directly with the theoretical prediction (green curve) and Fig. 6.39 of \cite{TS14}. The relative error between the numerically-determined value of $Ra_{cr}$ and the analytic solution are shown in the right panel of Fig \ref{fig:onset-1}.
```
Comment (bangerth) [inline doc/manual/manual.tex]: Write it as `Fig.~\ref{fig:onset-1}` (note the tilde) to keep the two parts on the same line.

same in the other two places in this line (the second one also misses the period after `Fig`).

File: doc/manual/manual.tex [hunk 15]
```diff
@@ -8168,7 +8168,53 @@ \subsubsection{Running benchmarks that require code}
...
+
+\begin{figure}
+ \includegraphics[width=0.49\textwidth]{cookbooks/benchmarks/onset-of-convection/racr.png}
+ \hfill
+ \includegraphics[width=0.49\textwidth]{cookbooks/benchmarks/onset-of-convection/racr_error.png}
+ \caption{(left) Comparison of numerically-determined and theoretical values for $Ra_{cr}$. Red circles indicate numerical simulations unstable to convection, black circles indicate simulations that are stable. The green dashed curve indicates the theoretical prediction. (right) Relative error in determination of $Ra_{cr}$. Dashed red line indicates error tolerance used in bisection procedure.}
```
Comment (bangerth) [inline doc/manual/manual.tex]: I think the typical style with other figures in the manual is to say `Left: ... . Right: ...`

**Comment by gassmoeller** [comment]
Hi Max,
does your commit address all of Wolfgang's comments? For us this is hard to judge, so leaving a comment after addressing comments helps us a lot in deciding if we can review/merge a pull request.

**Comment by maxrudolph** [comment]
Hi Rene,
Yes, I did my best to address all of Wolfgang's comments. I just re-read the manual text and fixed a couple of typos.

Max

**Comment by bangerth** [comment]
Let's go with it then!

**Pull request merged**